observability/logging.go

```diff
@@ -0,0 +1,310 @@
+	originalLogger iblog.Logger
+	// exporter is a loggingExporter and the handle for uploading collected data to backends
+	exporter unsafe.Pointer // loggingExporter
+	// filters is []*configpb.ObservabilityConfig_LogFilter used to set config on methods
+	filters unsafe.Pointer // []*configpb.ObservabilityConfig_LogFilter
+}
+
+func matchLogFilter(serviceMethod string, filters []*configpb.ObservabilityConfig_LogFilter) *configpb.ObservabilityConfig_LogFilter {
+	// Try matching the filters one by one, pick the first match. The
+	// correctness of the log filter pattern is ensured by config.go.
+	for _, filter := range filters {
+		if filter.Pattern == "*" {
+			// Match a "*"
+			return filter
+		}
+		tokens := strings.SplitN(filter.Pattern, "/", 2)
+		filterService := tokens[0]
+		filterMethod := tokens[1]
+		if filterMethod == "*" {
+			// Handle "p.s/*" case
+			s, _, err := grpcutil.ParseMethod(serviceMethod)
+			if err != nil {
+				// This service method string is provided by upstream code, if
+				// it's malformed, we are in bigger trouble.
+				logger.Errorf("Invalid service method: %v", err)
+				return nil
+			}
+			if s == filterService {
+				return filter
+			}
+			return nil
+		}
+		if serviceMethod == filter.Pattern {
+			// Exact match of "p.s/m"
+			return filter
+		}
+	}
+	return nil
+}
+
+func (l *observabilityBinaryLogger) GetMethodLogger(methodName string) iblog.MethodLogger {
+	var ol, ml iblog.MethodLogger
+
+	if l.originalLogger != nil {
+		ol = l.originalLogger.GetMethodLogger(methodName)
+	}
+
+	ePtr := atomic.LoadPointer(&l.exporter)
+	// If no exporter is specified, there is no point creating a method
+	// logger. We don't have any chance to inject exporter after its
+	// creation.
+	if ePtr == nil {
+		return ol
+	}
+	exporter := (*loggingExporter)(ePtr)
+
+	filtersPtr := atomic.LoadPointer(&l.filters)
+	if filtersPtr == nil {
+		// No filters set
+		return ol
+	}
+	filters := (*[]*configpb.ObservabilityConfig_LogFilter)(filtersPtr)
+
+	// If user specify a "*" pattern, binarylog will log every single call and
+	// content. This means the exporting RPC's events will be captured. Even if
+	// we batch up the uploads in the exporting RPC, the message content of that
+	// RPC will be logged. Without this exclusion, we may end up with an ever
+	// expanding message field in log entries, and crash the process with OOM.
+	if methodName == "google.logging.v2.LoggingServiceV2/WriteLogEntries" {
+		return ol
+	}
+
+	filter := matchLogFilter(methodName, *filters)
+	if filter == nil {
+		// No matching filter found
+		return ol
+	}
+	ml = iblog.NewMethodLogger(uint64(filter.HeaderBytes), uint64(filter.MessageBytes))
+	return &observabilityBinaryMethodLogger{
+		originalMethodLogger: ol,
+		wrappedMethodLogger:  ml,
+		rpcID:                uuid.NewString(),
+		exporter:             *exporter,
+	}
+}
+
+func (l *observabilityBinaryLogger) Close() {
+	if l == nil {
+		return
+	}
+	ePtr := atomic.LoadPointer(&l.exporter)
```

Comment:
I'd recommend a helper method for these atomic loads so the code that needs the pointers can be cleaner.